File: client/src/voice/sttDaemon.ts
// Daemon-backed STT.
//
// PCM16LE mono 16 kHz frames captured on the phone are sent as binary
// frames on the `ct-control` DataChannel. The daemon terminates xAI's
// streaming STT WebSocket server-side (Authorization header auth) and
// relays transcript events back as JSON control frames:
//
//   daemon → phone:
//     { t: "stt.ready" }
//     { t: "stt.partial", text, is_final }
//     { t: "stt.done",    text }
//     { t: "stt.error",   message }
//     { t: "stt.closed" }
//
//   phone → daemon:
//     { t: "stt.start" }
//     (binary PCM16LE frames)
//     { t: "stt.audio.done" }
//     { t: "stt.cancel" }
//
// No xAI key ever lives in the browser on this path.
//
// Audio capture is behind the `AudioSource` boundary from
// `./audioSource`. Production default is mic; `?audio-fixture=<url>`
// on the join URL switches to a deterministic fixture source that goes
// through this exact same daemon path.

import type { ControlMessage } from '../rtc/client';
import {
  MIC_BUFFER_SIZE,
  MicPermissionError,
  SAMPLE_RATE,
  selectAudioSource,
  type AudioSource,
} from './audioSource';
import { phoneToDaemon } from './protocol';

export { MicPermissionError } from './audioSource';

export interface SttChunkConfig {
  chunkMs: number;
  chunkBytes: number;
}

// Parses `?sttChunkMs=` debug param. Returns null for missing/invalid
// values so callers fall back to default per-frame send.
export function parseSttChunkMs(raw: string | null | undefined): SttChunkConfig | null {
  if (raw == null) return null;
  if (!/^\d+$/.test(raw)) return null;
  const ms = Number(raw);
  if (!Number.isFinite(ms)) return null;
  const chunkBytes = Math.round((ms * SAMPLE_RATE * 2) / 1000);
  return { chunkMs: ms, chunkBytes };
}

export function readSttChunkConfigFromLocation(): SttChunkConfig | null {
  if (typeof window === 'undefined') return null;
  const raw = new URLSearchParams(window.location.search).get('sttChunkMs');
  return parseSttChunkMs(raw);
}

export class DaemonNotConnectedError extends Error {
  constructor() {
    super('daemon_not_connected');
    this.name = 'DaemonNotConnectedError';
  }
}

export interface STTHandle {
  stop(): Promise<string>;
  cancel(): void;
}

export interface STTStartOptions {
  sendControl: (msg: ControlMessage) => void;
  sendBinary: (bytes: ArrayBuffer | Uint8Array) => void;
  addControlListener: (fn: (msg: ControlMessage) => void) => () => void;
  isConnected: () => boolean;
  onPartial?: (text: string, isFinal: boolean) => void;
  onError?: (reason: string) => void;
  onAudioFrame?: (pcm: ArrayBuffer) => void;
  // Override the audio source. Defaults to `selectAudioSource()` which
  // picks mic or fixture from the `?audio-fixture=` query param.
  audioSource?: AudioSource;
  // Debug-only batching of PCM frames before forwarding. When omitted,
  // each PCM frame is sent individually (the production default). When
  // provided, frames are accumulated until total bytes >= chunkBytes,
  // then flushed as a single binary message. Read from `?sttChunkMs=`
  // by default; pass explicitly in tests.
  chunkConfig?: SttChunkConfig | null;
}

// Rolling cap on pre-ready frames — bounded memory, enough to preserve
// the first ~1 s of mic audio captured while the daemon's xAI WS is
// still handshaking. Fixture source doesn't emit pre-ready so this is
// effectively unused there.
export const PRE_READY_CAP_FRAMES = Math.ceil(
  1000 / ((MIC_BUFFER_SIZE / SAMPLE_RATE) * 1000),
);

export async function startDaemonSTT(opts: STTStartOptions): Promise<STTHandle> {
  if (!opts.isConnected()) throw new DaemonNotConnectedError();

  const audioSource = opts.audioSource ?? selectAudioSource();

  let resolveReady!: () => void;
  let rejectReady!: (err: Error) => void;
  const ready = new Promise<void>((res, rej) => {
    resolveReady = res;
    rejectReady = rej;
  });

  let resolveFinal!: (text: string) => void;
  let rejectFinal!: (err: Error) => void;
  const finalTranscript = new Promise<string>((res, rej) => {
    resolveFinal = res;
    rejectFinal = rej;
  });

  let serverReady = false;
  let settled = false;

  const detach = opts.addControlListener((msg) => {
    if (msg.t === 'stt.ready') {
      if (!serverReady) {
        serverReady = true;
        resolveReady();
      }
      return;
    }
    if (msg.t === 'stt.partial') {
      const text = typeof msg.text === 'string' ? msg.text : '';
      const isFinal = !!(msg as { is_final?: boolean }).is_final;
      opts.onPartial?.(text, isFinal);
      return;
    }
    if (msg.t === 'stt.done') {
      if (!settled) {
        settled = true;
        resolveFinal(typeof msg.text === 'string' ? msg.text.trim() : '');
      }
      return;
    }
    if (msg.t === 'stt.error') {
      const reason = typeof msg.message === 'string' ? msg.message : 'stt_error';
      opts.onError?.(reason);
      if (!serverReady) rejectReady(new Error(reason));
      if (!settled) {
        settled = true;
        rejectFinal(new Error(reason));
      }
      return;
    }
    if (msg.t === 'stt.closed') {
      if (!serverReady) rejectReady(new Error('stt_closed_before_ready'));
      if (!settled) {
        settled = true;
        rejectFinal(new Error('stt_closed_before_done'));
      }
    }
  });

  const chunkConfig =
    opts.chunkConfig === undefined ? readSttChunkConfigFromLocation() : opts.chunkConfig;
  const batcher = chunkConfig ? createChunkBatcher(chunkConfig.chunkBytes, opts.sendBinary) : null;
  const sendOne = (pcm: ArrayBuffer) => {
    if (batcher) batcher.push(pcm);
    else opts.sendBinary(pcm);
  };

  let forwarding = false;
  const preReady: ArrayBuffer[] = [];
  const onFrame = (pcm: ArrayBuffer) => {
    opts.onAudioFrame?.(pcm);
    if (!forwarding) {
      preReady.push(pcm);
      while (preReady.length > PRE_READY_CAP_FRAMES) preReady.shift();
      return;
    }
    sendOne(pcm);
  };

  try {
    await audioSource.start(onFrame);
  } catch (err) {
    detach();
    throw err;
  }

  opts.sendControl(phoneToDaemon.sttStart());

  try {
    await ready;
  } catch (err) {
    await audioSource.stop();
    detach();
    opts.sendControl({ t: 'stt.cancel' });
    throw err instanceof Error ? err : new Error('stt_start_failed');
  }

  // Flush any frames captured between mic-arm and stt.ready (mic path),
  // then switch to live forwarding. Fixture path: preReady is empty
  // here, then `resume()` kicks off paced real-time emission that
  // arrives directly via the onFrame forwarding branch.
  for (const frame of preReady) sendOne(frame);
  preReady.length = 0;
  forwarding = true;
  audioSource.resume?.();

  const teardown = async () => {
    await audioSource.stop();
    detach();
  };

  return {
    async stop(): Promise<string> {
      batcher?.flush();
      opts.sendControl({ t: 'stt.audio.done' });
      try {
        return await finalTranscript;
      } finally {
        await teardown();
      }
    },
    cancel(): void {
      batcher?.discard();
      if (!settled) {
        settled = true;
        rejectFinal(new Error('stt_cancelled'));
      }
      opts.sendControl({ t: 'stt.cancel' });
      void teardown();
    },
  };
}

function createChunkBatcher(
  chunkBytes: number,
  sendBinary: (bytes: ArrayBuffer | Uint8Array) => void,
): { push(pcm: ArrayBuffer): void; flush(): void; discard(): void } {
  let pending: Uint8Array[] = [];
  let pendingBytes = 0;
  const flush = () => {
    if (pendingBytes === 0) {
      pending = [];
      return;
    }
    const out = new Uint8Array(pendingBytes);
    let offset = 0;
    for (const part of pending) {
      out.set(part, offset);
      offset += part.byteLength;
    }
    pending = [];
    pendingBytes = 0;
    sendBinary(out);
  };
  return {
    push(pcm) {
      const view = new Uint8Array(pcm);
      pending.push(view);
      pendingBytes += view.byteLength;
      if (pendingBytes >= chunkBytes) flush();
    },
    flush,
    discard() {
      pending = [];
      pendingBytes = 0;
    },
  };
}

File: android/app/src/main/java/app/clawkietalkie/voice/SttDaemon.kt
package app.clawkietalkie.voice

import app.clawkietalkie.protocol.ControlMessage
import app.clawkietalkie.protocol.PhoneToDaemon
import kotlinx.coroutines.CompletableDeferred
import kotlin.math.ceil

// Daemon-backed STT. Mirror of the web client's `src/voice/sttDaemon.ts`.
//
// PCM16LE mono 16 kHz frames captured on the phone are sent as binary frames
// on the DataChannel. The daemon records a turn, transcribes it through
// OpenClaw infer, and relays transcript events back as JSON control frames.
// No provider key ever lives on the phone on this path.

class DaemonNotConnectedError : Exception("daemon_not_connected")

interface SttHandle {
    /** Stops the mic, finalizes the turn, resolves with the final transcript. */
    suspend fun stop(): String
    fun cancel()
}

data class SttChunkConfig(val chunkMs: Int, val chunkBytes: Int)

/** Parses the `sttChunkMs` debug param (carried on handoff links). */
fun parseSttChunkMs(raw: String?): SttChunkConfig? {
    if (raw == null || !raw.matches(Regex("^\\d+$"))) return null
    val ms = raw.toIntOrNull() ?: return null
    val chunkBytes = Math.round(ms.toDouble() * SAMPLE_RATE * 2 / 1000).toInt()
    return SttChunkConfig(ms, chunkBytes)
}

class SttStartOptions(
    val sendControl: (ControlMessage) -> Unit,
    val sendBinary: (ByteArray) -> Unit,
    val addControlListener: ((ControlMessage) -> Unit) -> () -> Unit,
    val isConnected: () -> Boolean,
    val onPartial: ((text: String, isFinal: Boolean) -> Unit)? = null,
    val onError: ((reason: String) -> Unit)? = null,
    val onAudioFrame: ((pcm: ByteArray) -> Unit)? = null,
    val audioSource: AudioSource,
    // Debug-only batching of PCM frames before forwarding. When null, each
    // PCM frame is sent individually (the production default).
    val chunkConfig: SttChunkConfig? = null,
)

// Rolling cap on pre-ready frames — bounded memory, enough to preserve the
// first ~1 s of mic audio captured while the daemon STT session gets ready.
val PRE_READY_CAP_FRAMES = ceil(1000.0 / MIC_FRAME_DURATION_MS).toInt()

suspend fun startDaemonStt(opts: SttStartOptions): SttHandle {
    if (!opts.isConnected()) throw DaemonNotConnectedError()

    val audioSource = opts.audioSource

    val ready = CompletableDeferred<Unit>()
    val finalTranscript = CompletableDeferred<String>()

    var serverReady = false

    val detach = opts.addControlListener { msg ->
        when (msg.t) {
            "stt.ready" -> {
                if (!serverReady) {
                    serverReady = true
                    ready.complete(Unit)
                }
            }
            "stt.partial" -> {
                val text = msg.string("text") ?: ""
                val isFinal = msg.boolean("is_final") ?: false
                opts.onPartial?.invoke(text, isFinal)
            }
            "stt.done" -> {
                finalTranscript.complete((msg.string("text") ?: "").trim())
            }
            "stt.error" -> {
                val reason = msg.string("message") ?: "stt_error"
                opts.onError?.invoke(reason)
                if (!serverReady) ready.completeExceptionally(Exception(reason))
                finalTranscript.completeExceptionally(Exception(reason))
            }
            "stt.closed" -> {
                if (!serverReady) ready.completeExceptionally(Exception("stt_closed_before_ready"))
                finalTranscript.completeExceptionally(Exception("stt_closed_before_done"))
            }
        }
    }

    val batcher = opts.chunkConfig?.let { ChunkBatcher(it.chunkBytes, opts.sendBinary) }
    val sendOne: (ByteArray) -> Unit = { pcm ->
        if (batcher != null) batcher.push(pcm) else opts.sendBinary(pcm)
    }

    var forwarding = false
    val preReady = ArrayDeque<ByteArray>()
    val frameLock = Any()
    val onFrame: (ByteArray) -> Unit = { pcm ->
        opts.onAudioFrame?.invoke(pcm)
        synchronized(frameLock) {
            if (!forwarding) {
                preReady.addLast(pcm)
                while (preReady.size > PRE_READY_CAP_FRAMES) preReady.removeFirst()
            } else {
                sendOne(pcm)
            }
        }
    }

    try {
        // Source start may block (AudioRecord init, fixture fetch) — keep it
        // off the main thread.
        kotlinx.coroutines.withContext(kotlinx.coroutines.Dispatchers.IO) {
            audioSource.start(onFrame)
        }
    } catch (err: Exception) {
        detach()
        throw err
    }

    opts.sendControl(PhoneToDaemon.sttStart())

    try {
        ready.await()
    } catch (err: Exception) {
        audioSource.stop()
        detach()
        opts.sendControl(PhoneToDaemon.sttCancel())
        throw err
    }

    // Flush any frames captured between mic-arm and stt.ready, then switch to
    // live forwarding.
    synchronized(frameLock) {
        for (frame in preReady) sendOne(frame)
        preReady.clear()
        forwarding = true
    }
    audioSource.resume()

    val teardown = {
        audioSource.stop()
        detach()
    }

    return object : SttHandle {
        override suspend fun stop(): String {
            batcher?.flush()
            opts.sendControl(PhoneToDaemon.sttAudioDone())
            try {
                return finalTranscript.await()
            } finally {
                teardown()
            }
        }

        override fun cancel() {
            batcher?.discard()
            finalTranscript.completeExceptionally(Exception("stt_cancelled"))
            opts.sendControl(PhoneToDaemon.sttCancel())
            teardown()
        }
    }
}

internal class ChunkBatcher(
    private val chunkBytes: Int,
    private val sendBinary: (ByteArray) -> Unit,
) {
    private val pending = mutableListOf<ByteArray>()
    private var pendingBytes = 0

    @Synchronized
    fun push(pcm: ByteArray) {
        pending.add(pcm)
        pendingBytes += pcm.size
        if (pendingBytes >= chunkBytes) flushLocked()
    }

    @Synchronized
    fun flush() {
        flushLocked()
    }

    @Synchronized
    fun discard() {
        pending.clear()
        pendingBytes = 0
    }

    private fun flushLocked() {
        if (pendingBytes == 0) {
            pending.clear()
            return
        }
        val out = ByteArray(pendingBytes)
        var offset = 0
        for (part in pending) {
            part.copyInto(out, offset)
            offset += part.size
        }
        pending.clear()
        pendingBytes = 0
        sendBinary(out)
    }
}
